### ISSUE TITLE
What is actually being calculated in memory_efficient_attention?

### ISSUE THREAD

**Issue by henrycharlesworth** [facebookresearch/xformers#922]
In the docs of the memory_efficient_attention function it says:

```
:Equivalent pytorch code:

    .. code-block:: python

        scale = 1 / query.shape[-1] ** 0.5
        query = query * scale
        attn = query @ key.transpose(-2, -1)
        if attn_bias is not None:
            attn = attn + attn_bias
        attn = attn.softmax(-1)
        attn = F.dropout(attn, p)
        return attn @ value
```

But when I create a minimal example this doesn't seem to give the same result as what's actually calculated:

```
import torch

from xformers.ops import memory_efficient_attention

def equiv_fn(q, k, v, attn_bias=None):
    scale = 1.0 / q.shape[-1] ** 0.5
    q = q * scale
    attn = q @ k.transpose(-2, -1)
    if attn_bias is not None:
        attn = attn + attn_bias
    attn = attn.softmax(-1)
    return attn @ v


B = 16
T = 50
HD = 4
NH = 12
q = torch.randn(B, T, NH, HD, device="cuda")
k = torch.randn(B, T, NH, HD, device="cuda")
v = torch.randn(B, T, NH, HD, device="cuda")

out_1 = memory_efficient_attention(q, k, v)
out_2 = equiv_fn(q, k, v)

print(torch.allclose(out_1, out_2)) # False
print(torch.mean(torch.abs(out_1 - out_2))) # gives 0.355
```

I realise they probably shouldn't agree completely but they are not even close so I don't think it's numerical. I also tried passing `scale=1.0 / math.sqrt(HD)` into memory_efficient_attention and it still doesn't match (or get close).

So what is actually being calculated internally?

**Comment by henrycharlesworth** [comment]
OK, I was being a bit stupid, but the equivalent Pytorch code in the doc is wrong. It should be something like:

```
def equiv_fn(q, k, v, attn_bias=None):
    scale = 1.0 / q.shape[-1] ** 0.5
    q = q * scale
    q = q.transpose(1, 2)
    k = k.transpose(1, 2)
    v = v.transpose(1, 2)
    attn = q @ k.transpose(-2, -1)
    if attn_bias is not None:
        attn = attn + attn_bias
    attn = attn.softmax(-1)
    attn = attn @ v
    return attn.transpose(1, 2)
```

which then matches what memory_efficient_attention calculates.

**Comment by danthe3rd** [comment]
> the equivalent Pytorch code in the doc is wrong

Good catch, we should probably fix that. I believe the "equivalent" code in the doc only works for 3d inputs (with no head dimension)

**Comment by ajinkyavbhandare** [comment]
I would like to work on issue you mentioned . Could you assign me the same @danthe3rd  . Thankyou

**Comment by NielsRogge** [comment]
Hi,

For a model in the Transformers library I'd like to replace `xops.memory_efficient_attention` by native PyTorch, but it's only passing for 1e-4 absolute tolerance. 

The following will give an assertion error:

```python
import xformers.ops as xops
import torch

# batch_size, sequence_length, num_heads, head_dim)
queries = torch.randn(1, 8, 12, 512).to("cuda:0")
keys = torch.randn(1, 8, 12, 512).to("cuda:0")
values = torch.randn(1, 8, 12, 512).to("cuda:0")

scale = 1.0

out = xops.memory_efficient_attention(
    queries,
    keys,
    values,
    scale=scale,
)

# thanks to https://github.com/facebookresearch/xformers/issues/922#issuecomment-1808329588
queries = queries * scale
queries = queries.transpose(1, 2)
keys = keys.transpose(1, 2)
values = values.transpose(1, 2)
attention_scores = queries @ keys.transpose(-2, -1)

attention_probs = attention_scores.softmax(-1)
attention_output = attention_probs @ values
attention_output = attention_output.transpose(1, 2)

assert torch.allclose(out, attention_output, atol=1e-5)
```
I guess there's no way around this. It just slightly makes the logits differ

**Comment by fxmarty** [comment]
Any advice on this @danthe3rd, especially w.r.t fp16? Numerical differences between xformers/pytorch implem can accumulate quite a bit.

Trying a combination of https://github.com/pytorch/pytorch/pull/103167 & `torch.softmax(inp, dtype=torch.float32).to(torch.float16)` or simply `torch.softmax(inp)` does not help.
